fdbclient/include/fdbclient/KeyBackedTypes.actor.h

```diff
@@ -0,0 +1,1083 @@
+#endif
+
+	// Watch the trigger until it changes.  The result will be ready when it is observed that the trigger value's
+	// version is greater than initialVersion, and the observed trigger value version will be returned.
+	//
+	// If initialVersion is not present it will be initialized internally to the first read version successfully
+	// obtained from the db.
+	//
+	// initialVersion can be thought of as a "last known version" but it could also be used to indiciate some future
+	// version after which you want to be notifified if the trigger's value changes.
+	//
+	// If watching can be set, the initialized value of initialVersion will be sent to it once known.
+	template <class DB>
+	Future<Version> onChange(Reference<DB> db,
+	                         Optional<Version> initialVersion = {},
+	                         Promise<Version> watching = {}) const {
+		return onChangeActor(*this, db, initialVersion, watching);
+	}
+};
+
+ACTOR template <class DB>
+Future<Version> WatchableTrigger::onChangeActor(WatchableTrigger self,
+                                                Reference<DB> db,
+                                                Optional<Version> initialVersion,
+                                                Promise<Version> watching) {
+	state Reference<typename DB::TransactionT> tr = db->createTransaction();
+
+	loop {
+		if constexpr (can_set_transaction_options<DB>) {
+			db->setOptions(tr);
+		}
+		try {
+			// If the initialVersion is not set yet, then initialize it with the read version
+			if (!initialVersion.present()) {
+				wait(store(initialVersion, safeThreadFutureToFuture(tr->getReadVersion())));
+			}
+			if (watching.canBeSet()) {
+				watching.send(*initialVersion);
+			}
+
+			// Get the trigger's latest value.
+			Optional<Versionstamp> currentVal = wait(self.get(tr));
+
+			// If the trigger has a value and its version is > initialVersion then the trigger has fired so break
+			if (currentVal.present() && currentVal->version > *initialVersion) {
+				return currentVal->version;
+			}
+
+			// Otherwise, watch the key and repeat the loop once the watch fires
+			state Future<Void> watch = self.watch(tr);
+			wait(safeThreadFutureToFuture(tr->commit()));
+			wait(watch);
+			tr->reset();
+		} catch (Error& e) {
+			wait(safeThreadFutureToFuture(tr->onError(e)));
+		}
+	}
+}
+
+// Convenient read/write access to a single value of type T stored at key
+// Even though 'this' is not actually mutated, methods that change the db key are not const.
+template <typename T, typename Codec = TupleCodec<T>>
+class KeyBackedProperty {
+public:
+	KeyBackedProperty(KeyRef key = invalidKey, Optional<WatchableTrigger> trigger = {}, Codec codec = {})
+	  : key(key), trigger(trigger), codec(codec) {}
+
+	template <class Transaction>
+	Future<Optional<T>> get(Transaction tr, Snapshot snapshot = Snapshot::False) const {
+
+		if constexpr (is_transaction_creator<Transaction>) {
```

Comment:
My goal with the transaction creators is to extract them from these functions entirely and make it easy to apply them to all of the functionality for all of the types (including both reads and writes). This seems to work, though, for targeted application in functions.

Reply:
I agree with the goal but I'm not sure how to pull it off yet.  When we figure it out we can simplify these definitions, which is at least one step simpler than having two definitions of many functions.  There are of course other functions here that do not have this adaptation and do not already have two definitions.  I intended to add those later to round out the use cases here but only if we don't come up with some more clever way to do this.